src/ripple/rpc/impl/GRPCHelpers.cpp

```diff
@@ -0,0 +1,1854 @@
+// is called only if the field is present in the STObject
+template <class T>
+void
+populateProtoU8(STObject const& obj, SF_U8 const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->set_value(obj.getFieldU8(field));
+    }
+}
+
+template <class T>
+void
+populateProtoU16(STObject const& obj, SF_U16 const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->set_value(obj.getFieldU16(field));
+    }
+}
+
+template <class T>
+void
+populateProtoU32(STObject const& obj, SF_U32 const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->set_value(obj.getFieldU32(field));
+    }
+}
+
+template <class T>
+void
+populateProtoU64(STObject const& obj, SF_U64 const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->set_value(obj.getFieldU64(field));
+    }
+}
+
+template <class T>
+void
+populateProtoH128(STObject const& obj, SF_U128 const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->set_value(obj.getFieldH128(field).data(), uint128::size());
+    }
+}
+
+template <class T>
+void
+populateProtoH160(STObject const& obj, SF_U160 const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->set_value(obj.getFieldH160(field).data(), uint160::size());
+    }
+}
+
+template <class T>
+void
+populateProtoH256(STObject const& obj, SF_U256 const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->set_value(obj.getFieldH256(field).data(), uint256::size());
+    }
+}
+
+template <class T>
+void
+populateProtoVL(STObject const& obj, SF_Blob const& field, T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        auto data = obj.getFieldVL(field);
+        getProto()->set_value(data.data(), data.size());
+    }
+}
+
+template <class T>
+void
+populateProtoVLasString(
+    STObject const& obj,
+    SF_Blob const& field,
+    T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        auto data = obj.getFieldVL(field);
+        getProto()->set_value(
+            reinterpret_cast<const char*>(data.data()), data.size());
+    }
+}
+
+template <class T>
+void
+populateProtoVec256(
+    STObject const& obj,
+    SF_Vec256 const& field,
+    T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        const STVector256& vec = obj.getFieldV256(field);
+        for (size_t i = 0; i < vec.size(); ++i)
+        {
+            uint256 const& elt = vec[i];
+            getProto()->set_value(elt.data(), elt.size());
+        }
+    }
+}
+
+template <class T>
+void
+populateProtoAccount(
+    STObject const& obj,
+    SF_Account const& field,
+    T const& getProto)
+{
+    if (obj.isFieldPresent(field))
+    {
+        getProto()->mutable_value()->set_address(
+            toBase58(obj.getAccountID(field)));
+    }
+}
```

Comment:
What do you think about passing in a pointer to a member function and overloading these functions based on the return type of the member function. Something like:
```

template <class T>
void
populate(
    STObject const& from,
    T& to,
    org::xrpl::rpc::v1::Destination* (T::*get_mutable)())
{
    auto const& field = sfDestination;
    if (!from.isFieldPresent(field))
        return;

    (to.*get_mutable)()->mutable_value()->set_address(
        toBase58(from.getAccountID(field)));
}
```
It would be called like this:
```

template <class T>
void
populateDestination(STObject const& from, T& to)
{
    populate(from, to, &T::mutable_destination);
}
```

I _think_ this technique gets nicer if every field doesn't have its own type (although you have to pass in the field name in that case).

BTW, I don't object to using lambdas for the lazy evaluation, I think that's the right technique if we don't want to pass in the member pointer.

Reply:
I'll think about this some more, but I think passing in the member pointer is rather cryptic. For instance, I think a lot of people won't know what this snippet even means:
`org::xrpl::rpc::v1::Destination* (T::*get_mutable)())`
or what's going on here:
` (to.*get_mutable)()->`
You and I know what this means, but I don't think we are in the majority here.

Reply:
That's fair. It would be nice if I could overload based on a lambda return type, but I don't know how to do that (well, there's `std::function` or `enable_if`, but neither of those are a clear win).